nautobot/extras/models/models.py

```diff
@@ -701,12 +702,18 @@ def get_absolute_url(self):
         return reverse("extras:graphqlquery", kwargs={"slug": self.slug})
 
     def save(self, *args, **kwargs):
-        query_vars = re.findall(r"\$[a-zA-Z0-9]+", self.query)
-        vars_object = {}
-        for var in query_vars:
-            vars_object[var[1:]] = ""
-
-        self.variables = vars_object
+        variables = {}
+        schema = graphene_settings.SCHEMA
+        backend = get_default_backend()
+        document = backend.document_from_string(schema, self.query)
+        for definition in document.document_ast.definitions:
+            if isinstance(definition, OperationDefinition):
+                if definition.variable_definitions:
+                    for variable_definition in definition.variable_definitions:
+                        default = variable_definition.default_value.value if variable_definition.default_value else ""
+                        variables[variable_definition.variable.name.value] = default
```

Comment:
It's not immediately clear exactly what is going on here. Can you please add some commenting to this block?

Also since you're doing conditional iteration, perhaps pre-filtering to improve readability and reduce the need for so much nesting?  For example:

```suggestion
        definitions = [d for d in document.document_ast.definitions if isinstance(definition, OperationDefinition) and definition.variable_definitions]
        for definition in definitions:
            for variable_definition in definition.variable_definitions:
                default = variable_definition.default_value.value or ""  # boolean or is easier to read here
                variables[variable_definition.variable.name.value] = default
```

Reply:
This was based on my suggestion to introspect the parsed document tree to identify the variables, instead of trying to regex our way into finding them. I like your suggested changes though. Possible comment would be something like:

```
# Inspect the parsed document tree (document.document_ast) to retrieve the query (operation) definition(s)
# that define one or more variables. For each operation and variable definition, store the variable's
# default value (if any) into our own "variables" dict.
```